lib/actions/upsert.js

```diff
@@ -34,48 +32,36 @@ module.exports.objectTypes = async function getObjectTypes(configuration) {
  */
-module.exports.process = function upsertObject(message, configuration) {
-  const self = this;
+module.exports.process = async function upsertObject(message, configuration) {
   configuration.object = configuration.sobject;
-  console.log('Starting upsertObject');
-  const conn = new jsforce.Connection({
-    oauth2: {
-      clientId: process.env.OAUTH_CLIENT_ID,
-      clientSecret: process.env.OAUTH_CLIENT_SECRET,
-      redirectUri: 'https://app.elastic.io/oauth/v2',
-    },
-    instanceUrl: configuration.oauth.instance_url,
-    accessToken: configuration.oauth.access_token,
-    refreshToken: configuration.oauth.refresh_token,
-    version: SALESFORCE_API_VERSION,
-  });
-  conn.on('refresh', (accessToken, res) => {
-    console.log('Keys were updated, res=%j', res);
-    this.emit('updateKeys', {oauth: res});
-  });
+  this.logger.info('Starting upsertObject');
+
+  const conn = sfConnection.createConnection(configuration, this);
+
+  await attachment.prepareBinaryData(message, configuration, conn, this);
 
   if (message.body.Id) {
-    console.log('Upserting sobject=%s by internalId', configuration.sobject, message.body.Id);
-    debug('Upserting %s by internalId data: %j', configuration.sobject, message.body.Id, message);
+    this.logger.info('Upserting sobject=%s by internalId', configuration.sobject, message.body.Id);
+    this.logger.debug('Upserting %s by internalId data: %j', configuration.sobject, message.body.Id, message);
     return conn.sobject(configuration.sobject).update(message.body)
-        .then(() => {
-          configuration.lookupField = 'Id';
-          return lookup.process.call(self, {body:{Id: message.body.Id}},
-              configuration);
-        });
-  } else {
-    console.log('Upserting sobject: %s by externalId: %s',configuration.sobject, configuration.extIdField);
-    debug('Upserting sobject: %s by externalId:%s data: %j',configuration.sobject, configuration.extIdField, message);
+      .then(() => {
+        configuration.lookupField = 'Id';
+        return lookup.process.call(this, { body:{ Id: message.body.Id } },
+          configuration);
+      });
+  }
 
-    if(!configuration.extIdField){
-      throw Error('Can not find internalId/externalId ids');
-    }
+  this.logger.info('Upserting sobject: %s by externalId: %s',configuration.sobject, configuration.extIdField);
+  this.logger.debug('Upserting sobject: %s by externalId:%s data: %j', configuration.sobject, configuration.extIdField, message);
 
-    return conn.sobject(configuration.sobject)
-        .upsert(message.body, configuration.extIdField)
-        .then(() => {
-          configuration.lookupField = configuration.extIdField;
-          return lookup.process.call(this, message, configuration);
-        });
+  if (!configuration.extIdField) {
+    throw Error('Can not find internalId/externalId ids');
   }
+
+  return conn.sobject(configuration.sobject)
+    .upsert(message.body, configuration.extIdField)
+    .then(() => {
```

Comment:
Why it is not an `async` function?

Reply:
Here is a chain with the functions return promises. We don't need to wait at any point to get actual value.

  return conn.sobject(configuration.sobject)
    .upsert(message.body, configuration.extIdField)
    .then(() => {
      configuration.lookupField = configuration.extIdField;
      return lookup.process.call(this, message, configuration);
    });